### ISSUE TITLE
Overwrite existing file 

### ISSUE THREAD

**Issue by NadigNikhil** [hierynomus/smbj#106]
Hi,

I'm using the below code to overwrite an existing file. I'm getting the below exception. Kindly help me fix the issue.

Code:

```
SMBClient client = new SMBClient();
            Connection con = client.connect("ipAddress");
            AuthenticationContext context = new AuthenticationContext("username", "pwd".toCharArray(), "WORKGROUP");
            Session ses = con.authenticate(context);
            DiskShare share = (DiskShare) ses.connectShare("\\TEST_File");
            File file = share.openFile("TestFolder/TestFl1/1/1.txt", EnumSet.of(AccessMask.GENERIC_WRITE, AccessMask.MAXIMUM_ALLOWED), SMB2CreateDisposition.FILE_OVERWRITE);
            System.out.println(file.getFileId());
```

Exception:
com.hierynomus.smbj.common.SMBApiException: STATUS_SHARING_VIOLATION(3221225539/3221225539): Create failed for TestFolder/1.txt

**Comment by hierynomus** [comment]
Is the code really for that exception? The paths don't seem to match up...

2017-06-22 14:10 GMT+02:00 NadigNikhil <notifications@github.com>:

**Comment by pepijnve** [comment]
@hierynomus note that this is from an older version of smbj where the share mode could not yet be specified with `openFile`. On master you have to specify it.

@NadigNikhil a sharing violation means someone else has an open handle to the file and the specified  share mode does not allow you to open it as well. In 0.0.9, openFile always used `SMB2ShareAccess.FILE_SHARE_READ`. If someone else has the file in question open for write access for instance you'll get this type of error.
This area of the library has been made much more flexible on master, but there is no release available for this just yet.

**Comment by JhaShardendu** [comment]
@hierynomus that was a typo error ,the path is the same 

Exception:
com.hierynomus.smbj.common.SMBApiException: STATUS_SHARING_VIOLATION(3221225539/3221225539): Create failed for TestFolder/TestFl1/1/1.txt

**Comment by NadigNikhil** [comment]
@pepijnve Thanks for the update, but we need to give a feature in our product where the user can overwrite an existing file. When is the next release planned with the new feature?

**Comment by hierynomus** [comment]
Should not be too long. I aim for end of this month.



Op 23 jun. 2017 08:11 schreef "NadigNikhil" <notifications@github.com>:

**Comment by pepijnve** [comment]
@NadigNikhil if you need a workaround today you can 'inline' the openFile function to get access to the parameters you need.

The bit that needs changing is the `EnumSet.of(FILE_SHARE_READ)` parameter value. Change that to `EnumSet.of(FILE_SHARE_READ, FILE_SHARE_WRITE, FILE_SHARE_DELETE)`.

```
public File openFile(String path, EnumSet<AccessMask> accessMask, SMB2CreateDisposition createDisposition) throws TransportException, SMBApiException {
        logger.info("OpenFile {},{},{}", path, accessMask, createDisposition);

        SMB2FileId fileId = open(path, toLong(accessMask), null, EnumSet.of(FILE_SHARE_READ), createDisposition, EnumSet.of(SMB2CreateOptions.FILE_NON_DIRECTORY_FILE));
        // TODO
        return new File(fileId, treeConnect, path, 0L);
    }
```

**Comment by NadigNikhil** [comment]
@hierynomus and @pepijnve, Thanks for the update.

**Comment by JhaShardendu** [comment]
@hierynomus and @pepijnve,
We tried the above approach mentioned.It seems to work out at the moment. I have one more query is it possible to append the data to an existing file?

**Comment by pepijnve** [comment]
It's possible but rather cumbersome with 0.0.9. To do that you need to query the existing size of the file and then use File#write with a ByteChunkProvider with an initial file offset set to the size of the file.

On the development branch there are a bunch of additional methods on DiskShare and File to make this much more convenient.

**Comment by hierynomus** [comment]
Released 0.1.0 which contains the new methods. Closing this issue.
